Added Serialize and Deserialize traits

Techno-Fox opened the pull request:
serde is the most popular crate for serializing and de-serializing structs, enums, etc. I added them to the public facing data types so users can serialize and de-serialize the data returned from this crate.

Techno-Fox committed:
Added, and updated dependencies for serde capablilties

serde is the most used crate for serializing and de-serializing data (structs, enums, etc). Added dependencies for that

Techno-Fox committed:
Added Serialize and Deserialize traits 

Added Serialize and Deserialize traits to public structures. This allows user to use serde to serialize and deserialize them

Techno-Fox committed:
Added serde library

the serde library is the most used serializing and de-serializing library. This allows for users to serialize and de-serialize structures, enums, etc.

Techno-Fox committed:
Added serialize and deserialize for a PlatformCpuLoad

Forgot to added it for PlatformCpuLoad

unrelentingtech commented on Cargo.toml:
```diff
@@ -11,10 +11,13 @@ repository = "https://github.com/myfreeweb/systemstat"
-bytesize = "1.0"
+bytesize = { version = "1.0.1", features = ["serde"] }
 libc = "0.2"
+serde = {version = "1.0.126", features = ["derive"] }
```
Any reason to use at least that specific version rather than `1.0`? I generally specify only the major version.

Techno-Fox replied:
Sorry my vscode extension did that automatically.

unrelentingtech commented on Cargo.toml:
```diff
@@ -11,10 +11,13 @@ repository = "https://github.com/myfreeweb/systemstat"
+chrono = {version = "0.4.19", features = ["serde"] }
 lazy_static = "1.0"
-bytesize = "1.0"
+bytesize = { version = "1.0.1", features = ["serde"] }
```
Any reason to bump the version?

Techno-Fox replied:
Again sorry my vscode extension did that automatically.

unrelentingtech commented on Cargo.toml:
```diff
@@ -11,10 +11,13 @@ repository = "https://github.com/myfreeweb/systemstat"
 [dependencies]
 time = "0.1"
-chrono = "0.4"
+chrono = {version = "0.4.19", features = ["serde"] }
```
Missing space after the `{`. Any reason to bump the version?

Techno-Fox replied:
Again sorry my vscode extension did that automatically. I missed the space

unrelentingtech commented on Cargo.toml:
```diff
@@ -11,10 +11,13 @@ repository = "https://github.com/myfreeweb/systemstat"
 libc = "0.2"
+serde = {version = "1.0.126", features = ["derive"] }
+
+
```
Please don't add extra whitespace.

Techno-Fox replied:
Sorry for that.

valpackett reviewed (changes requested):
Would be nice if you could make this feature gated, like in the dependencies. I think that's how you do it with the `serde` feature name:

```rust
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
```

for the derives

```toml
[dependencies]
real_serde = { version = "1.0", package = "serde", features = ["derive"], optional = true }

[features]
serde = ["real_serde", "bytesize/serde", "chrono/serde"]
```

and something like that for the toml (and don't change the `chrono =` lines then), and use `extern crate real_serde as serde;` in `lib.rs`.

Techno-Fox commented:
I'm gonna need some help with this.

Here's how the lib.rs is looking like

```rs
//! This library provides a way to access system information such as CPU load, mounted filesystems,
//! network interfaces, etc.

#[cfg(windows)]
extern crate winapi;
extern crate libc;
extern crate time;
extern crate chrono;
extern crate bytesize;
#[cfg_attr(any(target_os = "freebsd", target_os = "openbsd", target_os = "macos"), macro_use)]
extern crate lazy_static;
#[cfg(any(target_os = "linux", target_os = "android"))]
extern crate nom;
#[cfg(feature = "serde")]
extern crate real_serde as serde;

pub mod platform;
pub mod data;

pub use self::platform::Platform;
pub use self::platform::PlatformImpl as System;
pub use self::data::*;
```

Cargo.toml : 
[package]
name = "systemstat"
version = "0.1.7"
authors = [ "Greg V <greg@unrelenting.technology>" ]
keywords = [ "System", "Info" ]
description = "systemstat"
license = "Unlicense"
readme = "README.md"
homepage = "https://github.com/myfreeweb/systemstat"
repository = "https://github.com/myfreeweb/systemstat"

[dependencies]
time = "0.1"
chrono = "0.4"
lazy_static = "1.0"
bytesize = "1.0"
libc = "0.2"
real_serde = { version = "1.0", package = "serde", features = ["derive"], optional = true }



[target.'cfg(any(target_os = "linux", target_os = "android"))'.dependencies]
nom = "6.0"

[target.'cfg(windows)'.dependencies.winapi]
version = "0.3"
features = ["fileapi", "sysinfoapi", "minwindef", "winbase", "winerror", "ws2def", "ws2ipdef", "pdh"]

[package.metadata.docs.rs]
targets = [
	"x86_64-unknown-freebsd",
	"x86_64-unknown-openbsd",
	"x86_64-unknown-linux-gnu",
	"x86_64-apple-darwin",
	"x86_64-pc-windows-msvc"
]

[features]
serde = ["real_serde", "bytesize/serde", "chrono/serde"]


I added #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
to the data types.

Now I'm getting a compilation error

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
                                                                        ^^^^^^^^^^^ can't find crate

unrelentingtech commented:
Hm, maybe https://crates.io/crates/serde-feature-hack actually is necessary.

Techno-Fox commented:
Same error happens (Sorry it took so long to comment back)

``rs
extern crate serde_feature_hack as serde;
```

```toml
[package]
name = "systemstat"
version = "0.1.7"
authors = [ "Greg V <greg@unrelenting.technology>" ]
keywords = [ "System", "Info" ]
description = "systemstat"
license = "Unlicense"
readme = "README.md"
homepage = "https://github.com/myfreeweb/systemstat"
repository = "https://github.com/myfreeweb/systemstat"

[dependencies]
time = "0.1"
chrono = "0.4"
lazy_static = "1.0"
bytesize = "1.0"
libc = "0.2"
serde-feature-hack = { version = "0.1.0", optional = true }


[target.'cfg(any(target_os = "linux", target_os = "android"))'.dependencies]
nom = "6.0"

[target.'cfg(windows)'.dependencies.winapi]
version = "0.3"
features = ["fileapi", "sysinfoapi", "minwindef", "winbase", "winerror", "ws2def", "ws2ipdef", "pdh"]

[package.metadata.docs.rs]
targets = [
	"x86_64-unknown-freebsd",
	"x86_64-unknown-openbsd",
	"x86_64-unknown-linux-gnu",
	"x86_64-apple-darwin",
	"x86_64-pc-windows-msvc"
]

[features]
serde = ["serde-feature-hack", "bytesize/serde", "chrono/serde"]
```

unrelentingtech commented:
Looks like the documentation there is outdated, you need version `0.2`, and then it would be just `extern crate serde` (no `as`!) and macros work fine.

Techno-Fox commented:
it works. I'll commit changes

Techno-Fox committed:
Added serde-feature-hack

added serde-feature-hack to add serde feautres

Techno-Fox committed:
added external crate serde

Added the external crate serde for macros

Techno-Fox committed:
Added serialize and deserialize as serde feature only

Added serialize and deserialize to be active only if the serde feature is activated

unrelentingtech commented:
Thanks!

willhughes11 commented:
New to Rust, trying to get the gated serde serializes to work, so I can serialize the platform objects but they don't.
Here is what my cargo.toml looks like.
`[package]
name = "events"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
systemstat = "0.2.2"
rusqlite = { version = "0.28.0", features = ["vtab", "csvtab", "modern_sqlite", "bundled"] }
json = "0.12.4"
log = "0.4.17"
env_logger = "0.9.3"
nanoid = "0.4.0"
serde = { version = "1.0.148", features = ["derive"], optional = true }
serde_json = "1.0.64"
regex = "1.7.0"

[features]
default = ["serde"]`

Error:
`the trait bound `Network: Serialize` is not satisfied
the following other types implement trait `Serialize`:`

valpackett commented:
@willhughes11 you haven't enabled the serde feature *in* systemstat, it should be 

```toml
default = ["serde", "systemstat/serde"]
```
